--- uboot.orig/env/sf.c	2022-02-02 20:03:18.035817804 +0530
+++ uboot/env/sf.c	2022-02-02 22:06:33.557635388 +0530
@@ -34,7 +34,9 @@
 #endif /* CONFIG_ENV_OFFSET_REDUND */
 
 DECLARE_GLOBAL_DATA_PTR;
-
+#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT 
+unsigned char broken_spi_banks = 0;// specify the number of broken SPI flash bank
+#endif
 static struct spi_flash *env_flash;
 
 static int setup_flash_device(void)
@@ -43,6 +45,20 @@
 	struct udevice *new;
 	int	ret;
 
+#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT 
+	unsigned int cs;
+	struct spi_flash *flash=NULL;
+    	for(cs=2;cs>0;cs--){
+		flash = spi_flash_probe(CONFIG_ENV_SPI_BUS, cs-1,CONFIG_ENV_SPI_MAX_HZ, CONFIG_ENV_SPI_MODE);		
+		if ((cs == 2) && (flash == NULL))
+		{
+			broken_spi_banks = 2;
+			continue;
+		}		    
+		spi_flash_free(flash);	  	
+    }    
+	
+#endif	
 	/* speed and mode will be read from DT */
 	ret = spi_flash_probe_bus_cs(CONFIG_ENV_SPI_BUS, CONFIG_ENV_SPI_CS,
 				     CONFIG_ENV_SPI_MAX_HZ, CONFIG_ENV_SPI_MODE,
--- uboot.orig/oem/ami/fmh/cmd_fmh.c	2022-02-02 20:03:22.275735063 +0530
+++ uboot/oem/ami/fmh/cmd_fmh.c	2022-02-02 20:14:03.604610961 +0530
@@ -71,7 +71,7 @@
         struct spi_flash *flashspi[CONFIG_SPX_FEATURE_GLOBAL_FLASH_BANKS]={NULL};
 #endif
 #ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT
-unsigned char broken_spi_banks = 0;// specify the number of broken SPI flash bank
+extern unsigned char broken_spi_banks; // specify the number of broken SPI flash bank
 #endif
 
 #ifdef CONFIG_BOOTLOGO_SUPPORT
