--- uboot.orig/oem/ami/ipmi/cmdhandler.c	2022-04-21 20:31:15.351130100 +0530
+++ uboot/oem/ami/ipmi/cmdhandler.c	2022-04-21 20:35:52.363622604 +0530
@@ -46,7 +46,7 @@
 
 extern int do_protect (cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[]);
 extern flash_info_t flash_info[];
-#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT 
+#if defined(CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT) && defined(CONFIG_SPX_FEATURE_DEDICATED_SPI_FLASH_BANK)
 extern unsigned char broken_spi_banks;// specify the number of broken SPI flash bank 
 #endif 
 
@@ -1302,7 +1302,7 @@
                 return sizeof(INT8U) + sizeof(pAMIDualImageSupRes->BootSelOpt.GetUploadSelector);
                 break;
 	case GETCURACTIVEIMG:
-#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT 
+#if defined(CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT) && defined(CONFIG_SPX_FEATURE_DEDICATED_SPI_FLASH_BANK)
                 if (broken_spi_banks == 1) 
                     pAMIDualImageSupRes->BootSelOpt.GetCurActiveImg = IMAGE_2_1BROKEN; 
                 else if (broken_spi_banks == 2) 
