--- uboot/arch/arm/mach-aspeed/ast2600/platform.S       2022-05-23 11:37:43.127906074 +0800
+++ uboot-new/arch/arm/mach-aspeed/ast2600/platform.S   2022-05-25 13:21:42.133215199 +0800
@@ -39,6 +39,7 @@
 #define AST_SCU_REV_ID                 (AST_SCU_BASE + 0x014)
 #define AST_SCU_SYSRST_CTRL            (AST_SCU_BASE + 0x040)
 #define AST_SCU_SYSRST_CTRL_CLR                (AST_SCU_BASE + 0x044)
+#define AST_SCU_SYSRST_EVENT           (AST_SCU_BASE + 0x064)
 #define AST_SCU_DEBUG_CTRL              (AST_SCU_BASE + 0x0C8)
 #define AST_SCU_DEBUG_CTRL2             (AST_SCU_BASE + 0x0D8)
 #define AST_SCU_HPLL_PARAM             (AST_SCU_BASE + 0x200)
@@ -237,13 +238,17 @@

 2:
        /* PCIeRC/E2M8 power-on reset comes from SCU040
-       But SCU04018 default value is 0x0.
-       It should be 0x1 to reset PCIeRC/E2M8.*/
+       It need set SCU040[18] high to reset PCIeRC/E2M
+       when AC power-on */
+       ldr     r0,  =AST_SCU_SYSRST_EVENT
+       ldr     r1, [r0]
+       tst     r1, #0x1
+       beq     3f
        ldr     r0, =AST_SCU_SYSRST_CTRL
        movw    r1, #0x0000
        movt    r1, #0x0004
        str     r1, [r0]
-
+3:
        /* Fix UART1 route problem on A3 */
        ldr     r0, =0x1e789098
        movw    r1, #0x0a30
--- uboot/drivers/clk/aspeed/clk_ast2600.c      2022-05-23 11:37:43.251906190 +0800
+++ uboot-new/drivers/clk/aspeed/clk_ast2600.c  2022-05-25 11:36:10.349198989 +0800
@@ -328,7 +328,7 @@
        u32 rate;
        u32 bclk_sel = (readl(&scu->clk_sel1) >> 20) & 0x7;

-       rate = ast2600_get_pll_rate(scu, ASPEED_CLK_HPLL);
+       rate = ast2600_get_pll_rate(scu, ASPEED_CLK_EPLL);

        return (rate / ((bclk_sel + 1) * 4));
 }
