diff -Naru lighttpd-1.4.64/src/Makefile.am ws/source/lighttpd_V1_4_64-13.0.0.0.0-src/data/.workspace/lighttpd/src/Makefile.am
--- lighttpd-1.4.64/src/Makefile.am	2022-01-19 23:23:51.000000000 +0530
+++ ws/source/lighttpd_V1_4_64-13.0.0.0.0-src/data/.workspace/lighttpd/src/Makefile.am	2022-07-14 17:43:46.250453082 +0530
@@ -1,17 +1,17 @@
 AM_CFLAGS = $(FAM_CFLAGS) $(LIBUNWIND_CFLAGS)
 
-noinst_PROGRAMS=\
-	t/test_common \
-	t/test_configfile \
-	t/test_mod
+#noinst_PROGRAMS=\
+#	t/test_common \
+#	t/test_configfile \
+#	t/test_mod
 
 sbin_PROGRAMS=lighttpd lighttpd-angel
 LEMON=$(top_builddir)/src/lemon$(BUILD_EXEEXT)
 
-TESTS=\
-	t/test_common$(EXEEXT) \
-	t/test_configfile$(EXEEXT) \
-	t/test_mod$(EXEEXT)
+#TESTS=\
+#	t/test_common$(EXEEXT) \
+#	t/test_configfile$(EXEEXT) \
+#	t/test_mod$(EXEEXT)
 
 lemon$(BUILD_EXEEXT): lemon.c
 	$(AM_V_CC)$(CC_FOR_BUILD) $(CPPFLAGS_FOR_BUILD) $(CFLAGS_FOR_BUILD) $(LDFLAGS_FOR_BUILD) -o $@ $(srcdir)/lemon.c
@@ -111,11 +111,11 @@
 mod_evasive_la_LDFLAGS = $(common_module_ldflags)
 mod_evasive_la_LIBADD = $(common_libadd)
 
-lib_LTLIBRARIES += mod_webdav.la
-mod_webdav_la_SOURCES = mod_webdav.c
-mod_webdav_la_CFLAGS = $(AM_CFLAGS) $(XML_CFLAGS) $(SQLITE_CFLAGS) 
-mod_webdav_la_LDFLAGS = $(common_module_ldflags)
-mod_webdav_la_LIBADD = $(common_libadd) $(XML_LIBS) $(SQLITE_LIBS) $(UUID_LIBS) $(ELFTC_LIB)
+#lib_LTLIBRARIES += mod_webdav.la
+#mod_webdav_la_SOURCES = mod_webdav.c
+#mod_webdav_la_CFLAGS = $(AM_CFLAGS) $(XML_CFLAGS) $(SQLITE_CFLAGS) 
+#mod_webdav_la_LDFLAGS = $(common_module_ldflags)
+#mod_webdav_la_LIBADD = $(common_libadd) $(XML_LIBS) $(SQLITE_LIBS) $(UUID_LIBS) $(ELFTC_LIB)
 
 if BUILD_WITH_LUA
 lib_LTLIBRARIES += mod_magnet.la
@@ -468,7 +468,7 @@
   mod_usertrack.c \
   mod_vhostdb.c \
   mod_vhostdb_api.c \
-  mod_webdav.c
+  #mod_webdav.c
 lighttpd_CPPFLAGS = \
   -DLIGHTTPD_STATIC \
   $(XML_CFLAGS) $(SQLITE_CFLAGS) \
