--- lighttpd-1.4.64/src/plugin_config.h	2022-01-19 23:23:51.000000000 +0530
+++ ws/source/lighttpd_V1_4_64-13.0.0.0.0-src/data/.workspace/lighttpd/src/plugin_config.h	2022-08-04 22:36:17.589258604 +0530
@@ -128,6 +128,14 @@
     uint8_t scope;   /* config_scope_type_t */
 } config_plugin_keys_t;
 
+typedef struct {
+	const char *key;
+	void *destination;
+
+	config_values_type_t type;
+	config_scope_type_t scope;
+} config_values_t;
+
 __attribute_cold__
 __attribute_pure__
 int config_plugin_value_tobool(const data_unset *du, int default_value);
