--- lighttpd-1.4.64/src/array.c	2022-01-19 23:23:51.000000000 +0530
+++ ws/source/lighttpd_V1_4_64-13.0.0.0.0-src/data/.workspace/lighttpd/src/array.c	2022-09-13 00:01:16.904675487 +0530
@@ -132,7 +132,29 @@
 
 	return a;
 }
+array *array_init_array(array *src) {
+	size_t i;
+	array *a = array_init(src->size);
 
+	if (0 == src->size) return a;
+
+	a->used = src->used;
+	a->size = src->size;
+	// a->unique_ndx = src->unique_ndx;
+
+	a->data = malloc(sizeof(*src->data) * src->size);
+	force_assert(NULL != a->data);
+	for (i = 0; i < src->size; i++) {
+        
+		if (src->data[i]) a->data[i] = src->data[i]->fn->copy(src->data[i]);
+		else a->data[i] = NULL;
+	}
+
+	a->sorted = malloc(sizeof(*src->sorted) * src->size);
+	force_assert(NULL != a->sorted);
+	memcpy(a->sorted, src->sorted, sizeof(*src->sorted) * src->size);
+	return a;
+}
 void array_free_data(array * const a) {
 	if (a->sorted) free(a->sorted);
 	data_unset ** const data = a->data;
