--- linux-5.4.85/drivers/char/Kconfig	2021-04-15 16:07:52.834764715 +0800
+++ linux-5.4.85-new/drivers/char/Kconfig	2021-04-19 10:59:13.469332638 +0800
@@ -16,6 +16,15 @@
 	  memory.
 	  When in doubt, say "Y".
 
+config DEVMEM_BOOTPARAM
+	bool "mem.devmem boot parameter"
+	depends on DEVMEM
+	default n
+	help
+	  This option adds a 'mem.devmem' kernel parameter which activates
+	  the /dev/mem device when enabled.
+	  When in doubt, say "N".
+
 config DEVKMEM
 	bool "/dev/kmem virtual device support"
 	# On arm64, VMALLOC_START < PAGE_OFFSET, which confuses kmem read/write
