--- linux-5.4.85/drivers/usb/gadget/udc/core.c	2021-04-15 16:08:43.625745311 +0800
+++ linux-5.4.85-new/drivers/usb/gadget/udc/core.c	2021-04-22 15:01:46.228096480 +0800
@@ -45,6 +45,7 @@
 static LIST_HEAD(udc_list);
 static LIST_HEAD(gadget_driver_pending_list);
 static DEFINE_MUTEX(udc_lock);
+static int virtual_device_status = 0;
 
 static int udc_bind_to_driver(struct usb_udc *udc,
 		struct usb_gadget_driver *driver);
@@ -1099,6 +1100,72 @@
 	udc->gadget->ops->udc_stop(udc->gadget);
 }
 
+int usb_gadget_vhub_upstream_enalbe(struct usb_gadget *gadget)
+{
+	struct usb_udc *udc;
+
+	if(virtual_device_status == 1)
+		return 0;
+	if(gadget != NULL) {
+		gadget->ops->udc_upstream_connect(gadget);
+	} else {
+	/* For now we take the first available UDC */
+		mutex_lock(&udc_lock);
+		list_for_each_entry(udc, &udc_list, list) {
+			if (udc->gadget!= NULL) {
+				break;
+			}
+		}
+		mutex_unlock(&udc_lock);
+		if(udc->gadget == NULL) {
+			//not found available gadget
+			return -1;
+		}
+		udc->gadget->ops->udc_upstream_connect(udc->gadget);	
+	}
+	virtual_device_status = 1;
+	return 0;
+}
+
+EXPORT_SYMBOL_GPL(usb_gadget_vhub_upstream_enalbe);
+
+int usb_gadget_vhub_upstream_disable(struct usb_gadget *gadget)
+{
+	struct usb_udc *udc;
+
+	if(virtual_device_status == 0)
+		return 0;
+	if(gadget != NULL) {
+		gadget->ops->udc_upstream_disconnect(gadget);
+	} else {
+	/* For now we take the first available UDC */
+		mutex_lock(&udc_lock);
+		list_for_each_entry(udc, &udc_list, list) {
+			if (udc->gadget!= NULL) {
+				break;
+			}
+		}
+		mutex_unlock(&udc_lock);
+		if(udc->gadget == NULL) {
+			//not found available gadget
+			return -1;
+		}
+		udc->gadget->ops->udc_upstream_disconnect(udc->gadget);	
+	}
+	virtual_device_status = 0;
+
+	return 0;
+}
+
+EXPORT_SYMBOL_GPL(usb_gadget_vhub_upstream_disable);
+
+int usb_gadget_get_all_device_status(int *status)
+{
+	*status = virtual_device_status;
+	return 0;
+}
+EXPORT_SYMBOL_GPL(usb_gadget_get_all_device_status);
+
 /**
  * usb_gadget_udc_set_speed - tells usb device controller speed supported by
  *    current driver
--- linux-5.4.85/include/linux/usb/gadget.h	2021-04-15 16:09:00.813399655 +0800
+++ linux-5.4.85-new/include/linux/usb/gadget.h	2021-04-22 15:03:00.050417696 +0800
@@ -322,6 +322,9 @@
 	struct usb_ep *(*match_ep)(struct usb_gadget *,
 			struct usb_endpoint_descriptor *,
 			struct usb_ss_ep_comp_descriptor *);
+	int (*udc_upstream_connect)(struct usb_gadget *gadget);
+	int (*udc_upstream_disconnect)(struct usb_gadget *gadget);
+
 };
 
 /**
@@ -850,6 +853,12 @@
 /* utility to tell udc core that the bus reset occurs */
 extern void usb_gadget_udc_reset(struct usb_gadget *gadget,
 		struct usb_gadget_driver *driver);
+/*-------------------------------------------------------------------------*/
+
+/* utility to set vhub connect/disconnect */
+extern int usb_gadget_vhub_upstream_enalbe(struct usb_gadget *gadget);
+extern int usb_gadget_vhub_upstream_disable(struct usb_gadget *gadget);
+extern int usb_gadget_get_all_device_status(int *status);
 
 /*-------------------------------------------------------------------------*/
 
