--- linux/drivers/usb/gadget/udc/Kconfig	2020-12-21 20:27:07.000000000 +0800
+++ linux.new/drivers/usb/gadget/udc/Kconfig	2021-04-13 14:02:32.068082443 +0800
@@ -441,6 +441,17 @@
 	  dynamically linked module called "udc-xilinx" and force all
 	  gadget drivers to also be dynamically linked.
 
+config USB_ASPEED
+        tristate "ASPEED USB Driver"
+        depends on HAS_DMA
+        depends on OF || COMPILE_TEST
+        help
+          USB peripheral controller driver for ASPEED USB device.
+
+          Say "y" to link the driver statically, or "m" to build a
+          dynamically linked module called "ast2600-udc" and force all
+          gadget drivers to also be dynamically linked.
+
 source "drivers/usb/gadget/udc/aspeed-vhub/Kconfig"
 
 #
