--- linux_old/drivers/spi/spi-aspeed.c	2022-03-07 15:55:33.810378338 +0800
+++ linux/drivers/spi/spi-aspeed.c	2022-03-07 16:11:04.262366509 +0800
@@ -1113,7 +1113,10 @@
 		ast_ctrl->chips[target_cs].max_clk_freq =
 			desc->mem->spi->max_speed_hz;
 
-		ret = info->calibrate(ast_ctrl, target_cs);
+		if (!of_property_read_bool(ast_ctrl->dev->of_node,
+			"timing-calibration-disabled")) {
+			ret = info->calibrate(ast_ctrl, target_cs);
+		}
 
 		dev_info(dev, "read bus width: %d [0x%08x]\n",
 			 op_tmpl.data.buswidth,
@@ -1208,7 +1211,7 @@
 }
 
 /* AST2600-A3 */
-static const struct spi_controller_mem_ops aspeed_spi_ops_cmd_read_cmd_write = {
+static const struct spi_controller_mem_ops aspeed_spi_ops_user_read_write = {
 	.exec_op = aspeed_spi_exec_op_user_mode,
 	.get_name = aspeed_spi_get_name,
 	.supports_op = aspeed_spi_support_op,
@@ -1427,7 +1430,7 @@
 		spi_ctrl->mem_ops = &aspeed_spi_ops_cmd_read_dma_write;
 	} else {
 		/* for AST2600-A3 */
-		spi_ctrl->mem_ops = &aspeed_spi_ops_cmd_read_cmd_write;
+		spi_ctrl->mem_ops = &aspeed_spi_ops_user_read_write;
 	}
 
 	spi_ctrl->dev.of_node = dev->of_node;
