--- b/drivers/pinctrl/aspeed/pinmux-aspeed.h	2023-08-07 15:05:22.347414356 +0530
+++ a/drivers/pinctrl/aspeed/pinmux-aspeed.h	2023-08-07 15:09:34.018610476 +0530
@@ -730,15 +730,6 @@
 			SIG_EXPR_LIST_PTR(pin, low), \
 			SIG_EXPR_LIST_PTR(pin, other))
 
-#define PIN_DECL_4(pin, other, prio1, prio2, prio3, prio4) \
-	SIG_EXPR_LIST_DECL_SESG(pin, other, other); \
-	PIN_DECL_(pin, \
-			SIG_EXPR_LIST_PTR(pin, prio1), \
-			SIG_EXPR_LIST_PTR(pin, prio2), \
-			SIG_EXPR_LIST_PTR(pin, prio3), \
-			SIG_EXPR_LIST_PTR(pin, prio4), \
-			SIG_EXPR_LIST_PTR(pin, other))
-
 #define GROUP_SYM(group) group_pins_ ## group
 #define GROUP_DECL(group, ...) \
 	static const int GROUP_SYM(group)[] = { __VA_ARGS__ }
