--- net-snmp-5.8_ori/include/net-snmp/library/snmpusm.h	2022-11-09 17:49:45.155473405 +0800
+++ net-snmp-5.8/include/net-snmp/library/snmpusm.h	2022-11-09 17:50:36.000000000 +0800
@@ -39,6 +39,13 @@
 
 #define USM_SEC_MODEL_NUMBER            SNMP_SEC_MODEL_USM
 
+#define AES_SIZE                  22
+#define MAX_ENC_HEX_PASSWORD_LEN  44
+#define TWO_CHAR_SIZE             3
+#define PREFIX_HEX_INDEX          0
+#define SUBFIX_HEX_INDEX          1
+#define PASSWORD_START_INDEX      2
+
     /*
      * Structures.
      */
